File: js/tilings/snub-trihexagonal.js
/**
 * Generates a Snub Trihexagonal tiling board (3.3.3.3.6).
 * This tiling consists of hexagons and triangles.
 * Every vertex is shared by one hexagon and four triangles.
 */
export function generateSnubTrihexagonalBoard(options) {
  const { cols, rows, tileSize: a, colorCount, rng } = options;

  const D = a * Math.sqrt(7);
  const alpha = Math.atan(1 / (3 * Math.sqrt(3)));

  const tiles = [];
  const hexMap = new Map();

  const getHexCenter = (q, r) => {
    const x = D * (q + r / 2);
    const y = D * (Math.sqrt(3) / 2) * r;
    return [x, y];
  };

  let idCounter = 0;
  const allVertices = [];

  // 1. Generate Hexagons
  for (let r = 0; r < rows; r++) {
    const r_offset = Math.floor(r / 2);
    for (let q = -r_offset; q < cols - r_offset; q++) {
      const [cx, cy] = getHexCenter(q, r);
      const id = idCounter++;

      const points = [];
      for (let i = 0; i < 6; i++) {
        const angle = alpha + (i * 60) * Math.PI / 180;
        const p = [cx + a * Math.cos(angle), cy + a * Math.sin(angle)];
        points.push(p);
        allVertices.push(p);
      }

      const tile = {
        id,
        colorId: Math.floor(rng() * colorCount),
        ownerId: null,
        points,
        neighbors: []
      };
      tiles.push(tile);
      hexMap.set(`${q},${r}`, id);
    }
  }

  // 2. Generate Triangles by finding equilateral triplets among hexagon vertices
  const distSq = (p1, p2) => {
    const dx = p1[0] - p2[0];
    const dy = p1[1] - p2[1];
    return dx * dx + dy * dy;
  };

  const targetSq = a * a;
  const tolerance = 0.1 * targetSq;

  // Spatial hash
  const grid = new Map();
  const getCellKey = (p) => `${Math.floor(p[0] / a)},${Math.floor(p[1] / a)}`;

  allVertices.forEach((v, idx) => {
    const key = getCellKey(v);
    if (!grid.has(key)) grid.set(key, []);
    grid.get(key).push(idx);
  });

  for (let i = 0; i < allVertices.length; i++) {
    const v1 = allVertices[i];
    const cx = Math.floor(v1[0] / a);
    const cy = Math.floor(v1[1] / a);

    for (let dx = -1; dx <= 1; dx++) {
      for (let dy = -1; dy <= 1; dy++) {
        const key2 = `${cx + dx},${cy + dy}`;
        const cell2 = grid.get(key2);
        if (!cell2) continue;

        for (const j of cell2) {
          if (j <= i) continue;
          const v2 = allVertices[j];
          const d12 = distSq(v1, v2);
          if (Math.abs(d12 - targetSq) < tolerance) {
            for (let dx2 = -1; dx2 <= 1; dx2++) {
              for (let dy2 = -1; dy2 <= 1; dy2++) {
                const key3 = `${cx + dx2},${cy + dy2}`;
                const cell3 = grid.get(key3);
                if (!cell3) continue;

                for (const k of cell3) {
                  if (k <= j) continue;
                  const v3 = allVertices[k];
                  const d13 = distSq(v1, v3);
                  const d23 = distSq(v2, v3);
                  if (Math.abs(d13 - targetSq) < tolerance && Math.abs(d23 - targetSq) < tolerance) {
                    tiles.push({
                      id: idCounter++,
                      colorId: Math.floor(rng() * colorCount),
                      ownerId: null,
                      points: [v1, v2, v3],
                      neighbors: []
                    });
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  // 3. Connectivity via proximity
  const addNeighbor = (idx1, id2) => {
    if (!tiles[idx1].neighbors.includes(id2)) {
      tiles[idx1].neighbors.push(id2);
    }
  };

  for (let i = 0; i < tiles.length; i++) {
    for (let j = i + 1; j < tiles.length; j++) {
      let common = 0;
      for (const p1 of tiles[i].points) {
        for (const p2 of tiles[j].points) {
          const dx = p1[0] - p2[0];
          const dy = p1[1] - p2[1];
          if (dx * dx + dy * dy < 0.01) {
            common++;
            break;
          }
        }
      }
      if (common >= 2) {
        addNeighbor(i, tiles[j].id);
        addNeighbor(j, tiles[i].id);
      }
    }
  }

  // 3. Finalize
  let minX = Infinity, minY = Infinity;
  let maxX = -Infinity, maxY = -Infinity;

  for (const t of tiles) {
    for (const p of t.points) {
      minX = Math.min(minX, p[0]);
      minY = Math.min(minY, p[1]);
      maxX = Math.max(maxX, p[0]);
      maxY = Math.max(maxY, p[1]);
    }
  }

  tiles.forEach(t => {
    t.points = t.points.map(p => [p[0] - minX, p[1] - minY]);
  });

  const startTileIds = [
    hexMap.get(`${0},${0}`),
    hexMap.get(`${cols - 1},0`),
    hexMap.get(`${-Math.floor((rows - 1) / 2)},${rows - 1}`),
    hexMap.get(`${cols - 1 - Math.floor((rows - 1) / 2)},${rows - 1}`)
  ].filter(id => id !== undefined);

  return {
    version: 1,
    generator: "snub-trihexagonal",
    width: maxX - minX,
    height: maxY - minY,
    tiles,
    startTileIds
  };
}

File: go-flood/pkg/tilings/snub_trihexagonal.go
package tilings

import (
	"fmt"
	"go-flood/pkg/core"
	"math"
)

func GenerateSnubTrihexagonalBoard(options Options) core.Board {
	a := options.TileSize
	D := a * math.Sqrt(7.0)
	alpha := math.Atan(math.Sqrt(3.0) / 5.0)
	cols, rows := options.Cols, options.Rows
	colorCount := options.ColorCount
	rng := options.RNG

	tiles := []core.Tile{}
	hexMap := make(map[string]int)

	getHexCenter := func(q, r int) (float64, float64) {
		x := D * (float64(q) + float64(r)/2.0)
		y := D * (math.Sqrt(3.0) / 2.0) * float64(r)
		return x, y
	}

	idCounter := 0
	type vertexInfo struct {
		id int
		p  core.Point
	}
	vertexMap := make(map[string]vertexInfo)
	getVertexId := func(p core.Point) int {
		key := fmt.Sprintf("%.3f,%.3f", p[0], p[1])
		if info, ok := vertexMap[key]; ok {
			return info.id
		}
		id := len(vertexMap)
		vertexMap[key] = vertexInfo{id, p}
		return id
	}

	// 1. Generate Hexagons
	type vTile struct {
		id   int
		vIds []int
	}
	var vTiles []vTile

	for r := 0; r < rows; r++ {
		rOffset := int(math.Floor(float64(r) / 2.0))
		for q := -rOffset; q < cols-rOffset; q++ {
			cx, cy := getHexCenter(q, r)
			id := idCounter
			idCounter++

			points := make([]core.Point, 6)
			vIds := make([]int, 6)
			for i := 0; i < 6; i++ {
				angle := alpha + float64(i*60)*math.Pi/180.0
				p := core.Point{cx + a*math.Cos(angle), cy + a*math.Sin(angle)}
				points[i] = p
				vIds[i] = getVertexId(p)
			}

			tiles = append(tiles, core.Tile{
				ID:        id,
				ColorID:   int(rng() * float64(colorCount)),
				OwnerID:   nil,
				Points:    points,
				Neighbors: []int{},
			})
			vTiles = append(vTiles, vTile{id, vIds})
			hexMap[fmt.Sprintf("%d,%d", q, r)] = id
		}
	}

	// 2. Generate Triangles
	var allVertices []core.Point
	verticesById := make([]core.Point, len(vertexMap))
	for _, info := range vertexMap {
		verticesById[info.id] = info.p
	}
	allVertices = verticesById

	distSq := func(p1, p2 core.Point) float64 {
		dx := p1[0] - p2[0]
		dy := p1[1] - p2[1]
		return dx*dx + dy*dy
	}

	targetSq := a * a
	tolerance := 0.1 * targetSq

	type cell struct{ x, y int }
	grid := make(map[cell][]int)
	getCell := func(p core.Point) cell {
		return cell{int(math.Floor(p[0] / a)), int(math.Floor(p[1] / a))}
	}
	for i, v := range allVertices {
		c := getCell(v)
		grid[c] = append(grid[c], i)
	}

	for i := 0; i < len(allVertices); i++ {
		v1 := allVertices[i]
		c1 := getCell(v1)

		for dx := -1; dx <= 1; dx++ {
			for dy := -1; dy <= 1; dy++ {
				c2 := cell{c1.x + dx, c1.y + dy}
				for _, j := range grid[c2] {
					if j <= i {
						continue
					}
					v2 := allVertices[j]
					if math.Abs(distSq(v1, v2)-targetSq) < tolerance {
						for dx2 := -1; dx2 <= 1; dx2++ {
							for dy2 := -1; dy2 <= 1; dy2++ {
								c3 := cell{c1.x + dx2, c1.y + dy2}
								for _, k := range grid[c3] {
									if k <= j {
										continue
									}
									v3 := allVertices[k]
									if math.Abs(distSq(v1, v3)-targetSq) < tolerance && math.Abs(distSq(v2, v3)-targetSq) < tolerance {
										id := idCounter
										idCounter++
										tiles = append(tiles, core.Tile{
											ID:        id,
											ColorID:   int(rng() * float64(colorCount)),
											OwnerID:   nil,
											Points:    []core.Point{v1, v2, v3},
											Neighbors: []int{},
										})
										vTiles = append(vTiles, vTile{id, []int{i, j, k}})
									}
								}
							}
						}
					}
				}
			}
		}
	}

	// 3. Optimized Connectivity
	vertexToTiles := make([][]int, len(allVertices))
	for tileIdx, vt := range vTiles {
		for _, vId := range vt.vIds {
			vertexToTiles[vId] = append(vertexToTiles[vId], tileIdx)
		}
	}

	for i := range tiles {
		neighborCounts := make(map[int]int)
		for _, vId := range vTiles[i].vIds {
			for _, j := range vertexToTiles[vId] {
				if i == j {
					continue
				}
				neighborCounts[j]++
			}
		}
		for j, count := range neighborCounts {
			if count >= 2 {
				tiles[i].Neighbors = append(tiles[i].Neighbors, tiles[j].ID)
			}
		}
	}

	// 4. Finalize
	minX, minY := math.MaxFloat64, math.MaxFloat64
	maxX, maxY := -math.MaxFloat64, -math.MaxFloat64
	for _, t := range tiles {
		for _, p := range t.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}

	for i := range tiles {
		for j := range tiles[i].Points {
			tiles[i].Points[j][0] -= minX
			tiles[i].Points[j][1] -= minY
		}
	}

	startTileIds := []int{}
	corners := []struct{ q, r int }{
		{0, 0},
		{cols - 1, 0},
		{-int(math.Floor(float64(rows-1) / 2.0)), rows - 1},
		{cols - 1 - int(math.Floor(float64(rows-1)/2.0)), rows - 1},
	}
	for _, c := range corners {
		if id, ok := hexMap[fmt.Sprintf("%d,%d", c.q, c.r)]; ok {
			startTileIds = append(startTileIds, id)
		}
	}

	return core.Board{
		Version:      1,
		Generator:    "snub-trihexagonal",
		Width:        maxX - minX,
		Height:       maxY - minY,
		Cols:         cols,
		Rows:         rows,
		Tiles:        tiles,
		StartTileIds: startTileIds,
	}
}
